src/components/EllipsisMiddle/index.tsx

```diff
@@ -0,0 +1,112 @@
+import { useResizeDetector } from 'react-resize-detector'
+import { FC, HTMLAttributes, useEffect, useMemo, useRef, useState } from 'react'
+import { assert } from '../../utils/error'
+import { createTextMeasurer } from '../../utils/string'
+
+const EllipsisMiddle: FC<
+  HTMLAttributes<HTMLDivElement> & {
+    children?: string
+    text?: string
+    minStartLen?: number
+    minEndLen?: number
+    onTruncateStateChange?: (truncated: boolean) => void
+    useTextWidthForPlaceholderWidth?: boolean
+  }
+> = ({
+  text,
+  children,
+  minStartLen = 0,
+  minEndLen = 0,
+  onTruncateStateChange,
+  useTextWidthForPlaceholderWidth,
+  ...divProps
+}) => {
+  const ref = useRef<HTMLDivElement>(null)
+  const { width: resizedWidth } = useResizeDetector({
+    targetRef: ref,
+    handleHeight: false,
+  })
+  const [parts, setParts] = useState<string[]>([''])
+  const [originTextWidth, setOriginTextWidth] = useState(0)
+
+  useEffect(() => {
+    if (!ref.current) return
+    const width = resizedWidth ?? ref.current.clientWidth
+
+    const fullText = text ?? children ?? ''
+    const measureText = createTextMeasurer(ref.current)
+
+    const fullWidth = measureText(fullText).width
+    setOriginTextWidth(fullWidth)
+    if (fullWidth <= width) {
+      setParts([fullText])
+      onTruncateStateChange?.(false)
+      return
+    }
+
+    const ellipsis = '...'
+    const remainingChars = Array.from(fullText)
+    const leftPart = remainingChars.splice(0, minStartLen)
+    const rightPart = remainingChars.splice(remainingChars.length - minEndLen, minEndLen)
+    let currentWidth = measureText([leftPart, ellipsis, rightPart].flat().join('')).width
+    let nextDirectionIsLeft = true
+
+    if (currentWidth > width) {
+      while (currentWidth > width) {
+        if (leftPart.length === 0) nextDirectionIsLeft = false
+        if (rightPart.length === 0 && !nextDirectionIsLeft) break
+
+        const char = nextDirectionIsLeft ? leftPart.pop() : rightPart.shift()
+        assert(char)
+        const charWidth = measureText(char).width
+        currentWidth -= charWidth
+        nextDirectionIsLeft = !nextDirectionIsLeft
+      }
+    } else {
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        const char = nextDirectionIsLeft ? remainingChars.shift() : remainingChars.pop()
+        assert(char)
+        const charWidth = measureText(char).width
+
+        if (currentWidth + charWidth > width) {
+          break
+        }
+
+        if (nextDirectionIsLeft) {
+          leftPart.push(char)
+        } else {
+          rightPart.unshift(char)
+        }
+        currentWidth += charWidth
+        nextDirectionIsLeft = !nextDirectionIsLeft
+      }
+    }
```

Comment:
Would it be kinda consuming? AFAIK, `ellipsis` in the middle is only used in addresses, while addresses are all mono font, it's easy to get the count of chars.

Reply:
No actual testing, I guess the performance overhead of the measureText call is not large, there is a perfomance overhead estimated in the createTextMeasurer, i will test to see.

EllipsisMiddle is a generic component, not just for use with mono font, so no such path optimization is done.

Reply:
Got it, it's fine if the performance is not obviously impacted

Reply:
Rule `object-curly-newline` removed, pushed